File: XRAiAssistant/XRAiAssistant/Library3D/Library3D.swift
import Foundation

// MARK: - 3D Library Protocol

protocol Library3D: Identifiable, Hashable {
    var id: String { get }
    var displayName: String { get }
    var description: String { get }
    var version: String { get }
    var playgroundTemplate: String { get }
    var systemPrompt: String { get }
    var defaultSceneCode: String { get }
    var codeLanguage: CodeLanguage { get }
    var iconName: String { get }
    var supportedFeatures: Set<Library3DFeature> { get }
    var documentationURL: String { get }
    var examples: [CodeExample] { get }
}

// MARK: - Code Example Model
struct CodeExample: Identifiable, Hashable {
    let id: String
    let title: String
    let description: String
    let code: String
    let category: ExampleCategory
    let difficulty: ExampleDifficulty

    // MARK: - Enhanced Metadata (for AI-assisted scene generation)
    let keywords: [String]           // Searchable keywords for finding similar examples
    let aiPromptHints: String?       // Hints for AI when generating similar scenes

    init(
        id: String? = nil,
        title: String,
        description: String,
        code: String,
        category: ExampleCategory,
        difficulty: ExampleDifficulty = .beginner,
        keywords: [String] = [],
        aiPromptHints: String? = nil
    ) {
        self.id = id ?? UUID().uuidString
        self.title = title
        self.description = description
        self.code = code
        self.category = category
        self.difficulty = difficulty
        self.keywords = keywords
        self.aiPromptHints = aiPromptHints
    }
}

enum ExampleCategory: String, CaseIterable {
    case basic = "Basic Shapes"
    case animation = "Animation"
    case lighting = "Lighting"
    case materials = "Materials"
    case interaction = "Interaction"
    case physics = "Physics"
    case effects = "Effects"
    case vr = "VR/XR"
    case advanced = "Advanced"

    var icon: String {
        switch self {
        case .basic: return "cube"
        case .animation: return "rotate.3d"
        case .lighting: return "lightbulb"
        case .materials: return "paintbrush"
        case .interaction: return "hand.tap"
        case .physics: return "figure.fall"
        case .effects: return "sparkles"
        case .vr: return "visionpro"
        case .advanced: return "gearshape.2"
        }
    }
}

enum ExampleDifficulty: String {
    case beginner = "Beginner"
    case intermediate = "Intermediate"
    case advanced = "Advanced"
}

// MARK: - Supporting Types

enum CodeLanguage: String, CaseIterable {
    case javascript = "javascript"
    case html = "html"
    case typescript = "typescript"
    
    var displayName: String {
        switch self {
        case .javascript: return "JavaScript"
        case .html: return "HTML"
        case .typescript: return "TypeScript"
        }
    }
    
    var monacoLanguage: String {
        switch self {
        case .javascript: return "javascript"
        case .html: return "html"
        case .typescript: return "typescript"
        }
    }
}

enum Library3DFeature: String, CaseIterable {
    case webgl = "webgl"
    case webxr = "webxr"
    case vr = "vr"
    case ar = "ar"
    case physics = "physics"
    case animation = "animation"
    case lighting = "lighting"
    case materials = "materials"
    case postProcessing = "postProcessing"
    case nodeEditor = "nodeEditor"
    case declarative = "declarative"
    case imperative = "imperative"
    
    var displayName: String {
        switch self {
        case .webgl: return "WebGL"
        case .webxr: return "WebXR"
        case .vr: return "VR Support"
        case .ar: return "AR Support"
        case .physics: return "Physics Engine"
        case .animation: return "Animation System"
        case .lighting: return "Advanced Lighting"
        case .materials: return "Material System"
        case .postProcessing: return "Post-Processing"
        case .nodeEditor: return "Node Editor"
        case .declarative: return "Declarative Syntax"
        case .imperative: return "Imperative Syntax"
        }
    }
}

// MARK: - Default Implementation

extension Library3D {
    // Default Hashable implementation
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
    
    // Default Equatable implementation
    static func == (lhs: Self, rhs: Self) -> Bool {
        return lhs.id == rhs.id
    }
}

// MARK: - Library3D Factory

enum Library3DFactory {
    static func createAllLibraries() -> [any Library3D] {
        return [
            BabylonJSLibrary(),
            ThreeJSLibrary(),
            AFrameLibrary(),
            ReactThreeFiberLibrary(),
            ReactylonLibrary()
        ]
    }
    
    static func createLibrary(id: String) -> (any Library3D)? {
        return createAllLibraries().first { $0.id == id }
    }
    
    static func getDefaultLibrary() -> any Library3D {
        return BabylonJSLibrary()
    }
}

File: XRAiAssistantAndroid/app/src/main/java/com/xrai/assistant/domain/model/Library3D.kt
package com.xrai.assistant.domain.model

/**
 * 3D Library interface defining the contract for 3D rendering frameworks
 *
 * iOS equivalent: Library3D protocol in Library3D.swift (line 5)
 *
 * Implementations: Babylon.js, Three.js, A-Frame, React Three Fiber, Reactylon
 */
interface Library3D {
    val id: String
    val displayName: String
    val description: String
    val version: String
    val playgroundTemplate: String
    val systemPrompt: String
    val defaultSceneCode: String
    val codeLanguage: CodeLanguage
    val iconName: String
    val supportedFeatures: Set<Library3DFeature>
    val documentationURL: String
    val examples: List<CodeExample>
}

/**
 * Code example with metadata for AI-assisted learning
 *
 * iOS equivalent: CodeExample struct in Library3D.swift (line 21)
 */
data class CodeExample(
    val id: String = java.util.UUID.randomUUID().toString(),
    val title: String,
    val description: String,
    val code: String,
    val category: ExampleCategory,
    val difficulty: ExampleDifficulty = ExampleDifficulty.BEGINNER,
    val keywords: List<String> = emptyList(),
    val aiPromptHints: String? = null
)

/**
 * Example categories
 *
 * iOS equivalent: ExampleCategory enum in Library3D.swift (line 54)
 */
enum class ExampleCategory(
    val displayName: String,
    val icon: String
) {
    BASIC("Basic Shapes", "cube"),
    ANIMATION("Animation", "rotate_3d"),
    LIGHTING("Lighting", "lightbulb"),
    MATERIALS("Materials", "palette"),
    INTERACTION("Interaction", "touch_app"),
    PHYSICS("Physics", "sports_handball"),
    EFFECTS("Effects", "auto_awesome"),
    VR("VR/XR", "view_in_ar"),
    ADVANCED("Advanced", "settings")
}

/**
 * Example difficulty levels
 *
 * iOS equivalent: ExampleDifficulty enum in Library3D.swift (line 80)
 */
enum class ExampleDifficulty(val displayName: String) {
    BEGINNER("Beginner"),
    INTERMEDIATE("Intermediate"),
    ADVANCED("Advanced")
}

/**
 * Code language types
 *
 * iOS equivalent: CodeLanguage enum in Library3D.swift (line 88)
 */
enum class CodeLanguage(
    val rawValue: String,
    val displayName: String,
    val monacoLanguage: String
) {
    JAVASCRIPT("javascript", "JavaScript", "javascript"),
    HTML("html", "HTML", "html"),
    TYPESCRIPT("typescript", "TypeScript", "typescript")
}

/**
 * Library features
 *
 * iOS equivalent: Library3DFeature enum in Library3D.swift (line 110)
 */
enum class Library3DFeature(val displayName: String) {
    WEBGL("WebGL"),
    WEBXR("WebXR"),
    VR("VR Support"),
    AR("AR Support"),
    PHYSICS("Physics Engine"),
    ANIMATION("Animation System"),
    LIGHTING("Advanced Lighting"),
    MATERIALS("Material System"),
    POST_PROCESSING("Post-Processing"),
    NODE_EDITOR("Node Editor"),
    DECLARATIVE("Declarative Syntax"),
    IMPERATIVE("Imperative Syntax")
}

/**
 * Factory for creating library instances
 *
 * iOS equivalent: Library3DFactory enum in Library3D.swift (line 158)
 */
object Library3DFactory {
    /**
     * Create all available libraries
     */
    fun createAllLibraries(): List<Library3D> {
        return listOf(
            BabylonJSLibrary(),
            ThreeJSLibrary(),
            AFrameLibrary(),
            ReactThreeFiberLibrary(),
            ReactylonLibrary()
        )
    }

    /**
     * Create library by ID
     */
    fun createLibrary(id: String): Library3D? {
        return createAllLibraries().find { it.id == id }
    }

    /**
     * Get default library (Babylon.js)
     */
    fun getDefaultLibrary(): Library3D {
        return BabylonJSLibrary()
    }
}

/**
 * Placeholder implementations (will be implemented in separate files)
 */
private class BabylonJSLibrary : Library3D {
    override val id = "babylonjs"
    override val displayName = "Babylon.js"
    override val description = "Powerful WebGL engine for 3D rendering"
    override val version = "6.0+"
    override val playgroundTemplate = ""
    override val systemPrompt = ""
    override val defaultSceneCode = ""
    override val codeLanguage = CodeLanguage.JAVASCRIPT
    override val iconName = "view_in_ar"
    override val supportedFeatures = setOf(
        Library3DFeature.WEBGL,
        Library3DFeature.WEBXR,
        Library3DFeature.VR,
        Library3DFeature.AR,
        Library3DFeature.PHYSICS,
        Library3DFeature.ANIMATION,
        Library3DFeature.LIGHTING,
        Library3DFeature.MATERIALS,
        Library3DFeature.POST_PROCESSING,
        Library3DFeature.NODE_EDITOR,
        Library3DFeature.IMPERATIVE
    )
    override val documentationURL = "https://doc.babylonjs.com/"
    override val examples = emptyList<CodeExample>()
}

private class ThreeJSLibrary : Library3D {
    override val id = "threejs"
    override val displayName = "Three.js"
    override val description = "Popular JavaScript 3D library"
    override val version = "0.171.0"
    override val playgroundTemplate = ""
    override val systemPrompt = ""
    override val defaultSceneCode = ""
    override val codeLanguage = CodeLanguage.JAVASCRIPT
    override val iconName = "category"
    override val supportedFeatures = setOf(
        Library3DFeature.WEBGL,
        Library3DFeature.WEBXR,
        Library3DFeature.VR,
        Library3DFeature.AR,
        Library3DFeature.ANIMATION,
        Library3DFeature.LIGHTING,
        Library3DFeature.MATERIALS,
        Library3DFeature.POST_PROCESSING,
        Library3DFeature.IMPERATIVE
    )
    override val documentationURL = "https://threejs.org/docs/"
    override val examples = emptyList<CodeExample>()
}

private class AFrameLibrary : Library3D {
    override val id = "aframe"
    override val displayName = "A-Frame"
    override val description = "Web framework for building VR experiences"
    override val version = "1.4.0"
    override val playgroundTemplate = ""
    override val systemPrompt = ""
    override val defaultSceneCode = ""
    override val codeLanguage = CodeLanguage.HTML
    override val iconName = "view_in_ar"
    override val supportedFeatures = setOf(
        Library3DFeature.WEBGL,
        Library3DFeature.WEBXR,
        Library3DFeature.VR,
        Library3DFeature.AR,
        Library3DFeature.DECLARATIVE
    )
    override val documentationURL = "https://aframe.io/docs/"
    override val examples = emptyList<CodeExample>()
}

private class ReactThreeFiberLibrary : Library3D {
    override val id = "reactThreeFiber"
    override val displayName = "React Three Fiber"
    override val description = "React renderer for Three.js"
    override val version = "8.17.10"
    override val playgroundTemplate = ""
    override val systemPrompt = ""
    override val defaultSceneCode = ""
    override val codeLanguage = CodeLanguage.TYPESCRIPT
    override val iconName = "code"
    override val supportedFeatures = setOf(
        Library3DFeature.WEBGL,
        Library3DFeature.WEBXR,
        Library3DFeature.VR,
        Library3DFeature.AR,
        Library3DFeature.DECLARATIVE
    )
    override val documentationURL = "https://docs.pmnd.rs/react-three-fiber/"
    override val examples = emptyList<CodeExample>()
}

private class ReactylonLibrary : Library3D {
    override val id = "reactylon"
    override val displayName = "Reactylon"
    override val description = "React renderer for Babylon.js"
    override val version = "3.2.1"
    override val playgroundTemplate = ""
    override val systemPrompt = ""
    override val defaultSceneCode = ""
    override val codeLanguage = CodeLanguage.TYPESCRIPT
    override val iconName = "code"
    override val supportedFeatures = setOf(
        Library3DFeature.WEBGL,
        Library3DFeature.WEBXR,
        Library3DFeature.VR,
        Library3DFeature.AR,
        Library3DFeature.DECLARATIVE
    )
    override val documentationURL = "https://www.reactylon.com/"
    override val examples = emptyList<CodeExample>()
}
